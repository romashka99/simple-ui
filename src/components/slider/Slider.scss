@import '../../styles/colors.scss';
@import '../../styles/main.scss';
@import '../../styles/style.scss';

.slider {
	display: flex;
	position: relative;
	cursor: pointer;

	padding: 10px;

	.slider--inner {
		position: relative;
		height: 5px;
		width: 100%;
		background-color: $color-mid-light-grey;
		border-radius: 5px;
	}

	.slider--filled {
		position: absolute;
		height: 5px;
		border-radius: 5px;
		background-color: $color-grey;
	}

	.slider--control {
		cursor: pointer;
		position: absolute;
		height: 23px;
		width: 23px;
		border: $border;
		border-radius: 50%;
		background-color: $color-white;
		right: 0;
		top: -10px;
	}
}
