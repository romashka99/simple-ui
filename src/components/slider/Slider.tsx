import React, { useCallback, useRef, useState } from 'react';
import './Slider.scss';

interface ISliderProps extends React.AriaAttributes {
	min?: number;
	max?: number;
	value?: number;
	step?: number;
	defaultValue?: number;
	disabled?: boolean;
	className?: string;
	onChange?: (value: number) => void;
}

const range = (start: number, stop: number, step: number) =>
	Array.from({ length: (stop - start) / step + 1 }, (_, i) => start + i * step);

export default function Slider({
	min = 0,
	max = 100,
	step = 10,
	value,
	defaultValue = 0,
	disabled,
	className,
	onChange,
}: ISliderProps) {
	const classes = `slider${disabled ? ' is-disabled' : ''}${
		className ? ' ' + className : ''
	}`;

	const defaultValueSlider = Number(value ? value : defaultValue);

	const [valueSlider, setValueSlider] = useState<number>(defaultValueSlider);
	const [clicked, setClicked] = useState<boolean>(false);

	const values = range(min, max, step);

	const percent = 100 / values.length;

	const refInner = useRef<HTMLDivElement>(null);

	const onToggle = useCallback(
		(event: React.MouseEvent<HTMLElement>) => {
			const width = refInner.current ? refInner.current.offsetWidth : 0;
			const { clientX } = event;
			const percents = (clientX * 100) / width;
			const index = Math.round(percents / percent);
			const changeValue = values[index - 1];
			if (changeValue !== valueSlider) {
				setValueSlider(values[index - 1]);
				if (onChange) onChange(values[index - 1]);
			}
		},
		[onChange, percent, valueSlider, values]
	);

	const onClick = useCallback(
		(event: React.MouseEvent<HTMLElement>) => {
			if (clicked) {
				onToggle(event);
			}
		},
		[clicked, onToggle]
	);

	return (
		<div data-testId='slider' className={classes} onClick={onToggle}>
			<div
				ref={refInner}
				data-testId='slider-inner'
				className='slider--inner'
				onMouseMove={onClick}
				onMouseDown={() => setClicked(true)}
				onMouseUp={() => setClicked(false)}
			>
				<div
					style={{ width: `${valueSlider * percent}%` }}
					data-testId='slider-filled'
					className='slider--filled'
				>
					<button
						data-testId='slider-control'
						className='slider--control'
					></button>
				</div>
			</div>
		</div>
	);
}
