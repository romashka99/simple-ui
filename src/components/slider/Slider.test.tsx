import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Slider from '.';

describe('Test Slider', () => {
	test('without attr', () => {
		render(<Slider />);
		const slider = screen.getByTestId('slider');
		const inner: HTMLDivElement = within(slider).getByTestId('slider-inner');
		const filled: HTMLDivElement = within(slider).getByTestId('slider-filled');
		const control: HTMLButtonElement =
			within(slider).getByTestId('slider-control');
		expect(inner).not.toBeNull();
		expect(filled).not.toBeNull();
		expect(control).not.toBeNull();
		expect(slider.classList.length).toBe(1);
		expect(slider.classList.contains('slider')).toBe(true);
		expect(slider).toMatchSnapshot();
	});

	test('disabled', () => {
		render(<Slider disabled />);
		const slider = screen.getByTestId('slider');
		expect(slider.classList.length).toBe(2);
		expect(slider.classList.contains('slider')).toBe(true);
		expect(slider.classList.contains('is-disabled')).toBe(true);
		expect(slider).toMatchSnapshot();
	});

	test('onChange', () => {
		let value = 0;
		render(
			<Slider
				onChange={() => {
					value = 1;
				}}
			/>
		);
		const slider = screen.getByTestId('slider');
		const inner: HTMLDivElement = within(slider).getByTestId('slider-inner');
		expect(value).toBe(0);
		userEvent.click(inner);
		expect(value).toBe(1);
		expect(slider).toMatchSnapshot();
	});
});
