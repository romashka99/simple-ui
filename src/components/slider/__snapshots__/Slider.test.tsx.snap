// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Slider disabled 1`] = `
<div
  class="slider is-disabled"
  data-testid="slider"
>
  <div
    class="slider--inner"
    data-testid="slider-inner"
  >
    <div
      class="slider--filled"
      data-testid="slider-filled"
      style="width: 0%;"
    >
      <button
        class="slider--control"
        data-testid="slider-control"
      />
    </div>
  </div>
</div>
`;

exports[`Test Slider onChange 1`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider--inner"
    data-testid="slider-inner"
  >
    <div
      class="slider--filled"
      data-testid="slider-filled"
      style="width: 0%;"
    >
      <button
        class="slider--control"
        data-testid="slider-control"
      />
    </div>
  </div>
</div>
`;

exports[`Test Slider without attr 1`] = `
<div
  class="slider"
  data-testid="slider"
>
  <div
    class="slider--inner"
    data-testid="slider-inner"
  >
    <div
      class="slider--filled"
      data-testid="slider-filled"
      style="width: 0%;"
    >
      <button
        class="slider--control"
        data-testid="slider-control"
      />
    </div>
  </div>
</div>
`;
